dnfdaemon-server/main.cpp

```diff
@@ -22,17 +22,40 @@ along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
 #include "session_manager.hpp"
 
 #include <sdbus-c++/sdbus-c++.h>
+#include <signal.h>
 
 #include <iostream>
 
-static std::unique_ptr<SessionManager> session_manager = nullptr;
+static std::mutex session_manager_ptr_mutex;
+static std::atomic<SessionManager *> session_manager_ptr{nullptr};
+
+void sigint_handler([[maybe_unused]] int signum) {
+    // make sure that session_manager.shut_down() is called only once
+    SessionManager * local_session_manager_ptr{nullptr};
+    {
+        std::lock_guard<std::mutex> lock(session_manager_ptr_mutex);
+        local_session_manager_ptr = session_manager_ptr.load();
+        session_manager_ptr = nullptr;
+    }
+    if (local_session_manager_ptr != nullptr) {
+        local_session_manager_ptr->shut_down();
+    }
+}
 
 int main() {
+    struct sigaction signal_action;
+    memset(&signal_action, 0, sizeof(signal_action));
+    signal_action.sa_handler = sigint_handler;
+    sigaction(SIGINT, &signal_action, NULL);
+    sigaction(SIGTERM, &signal_action, NULL);
+
     try {
-        session_manager = std::make_unique<SessionManager>();
```

Comment:
You should also lock `session_manager_ptr_mutex` here, as a race could also occur between the handler and creating the session manager.